test/unit/math/mix/prob/categorical_logit_test.cpp

```diff
@@ -4,16 +4,14 @@
 #include <vector>
 
 TEST(ProbDistributionsCategoricalLogit, fvar_var) {
-  using Eigen::Dynamic;
-  using Eigen::Matrix;
-  using stan::math::fvar;
   using stan::math::log_softmax;
-  using stan::math::var;
-  Matrix<fvar<var>, Dynamic, 1> theta(3, 1);
+  using stan::math::vector_fv;
```

Comment:
Let's just replace these fwd/mix tests for all these distributions with `expect_ad` ones. I think it's possible and it'll make this simpler to work with in the future.

For instance, something like this here.

```cpp
TEST(ProbDistributionsCategoricalLogit, expect_ad) {
  auto f = [](const auto& y) {
    [](const auto& beta) {
      return stan::math::categorical_logit_lpdf(y, beta);
    };
  };

  int y1 = 1;
  int y2 = 2;
  Eigen::VectorXd beta1(1);
  beta1 << 1.0;
  Eigen::VectorXd beta2(2);
  beta2 << -0.1, 0.2;
  stan::test::expect_ad(f(y1), beta1);
  stan::test::expect_ad(f(y2), beta2);

  std::vector<int> y1s = { y1, y1 };
  std::vector<int> y2s = { y1, y2 };

  std::vector<Eigen::VectorXd> beta1s = { beta1, 2.0 * beta1 };
  std::vector<Eigen::VectorXd> beta2s = { beta2, 2.0 * beta2 };

  stan::test::expect_ad(f(y1s), beta1);
  stan::test::expect_ad(f(y1), beta1s);
  stan::test::expect_ad(f(y1s), beta1s);
  stan::test::expect_ad(f(y2s), beta2);
  stan::test::expect_ad(f(y2), beta2s);
  stan::test::expect_ad(f(y2s), beta2s);

  // Expect error
  stan::test::expect_ad(f(y2s), beta1s);
```

Reply:
Using the AD testing framework seems to be tripping up the simplex check due to some floating-point-ness:
Using the test:
```
#include <test/unit/math/test_ad.hpp>

TEST(ProbDistributionsCategorical, mix_t) {
  auto f = [](const auto& x1) {
    return [=](const auto& x2) { return stan::math::categorical_log(x1, x2); };
  };

  Eigen::VectorXd theta(3);
  theta << 0.2, 0.2, 0.6;

  stan::test::expect_ad(f(1), theta);
}
```

Gives the error:
```
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from ProbDistributionsCategorical
[ RUN      ] ProbDistributionsCategorical.mix_t
unknown file: Failure
C++ exception with description "categorical_lpmf: Probabilities parameter is not a valid simplex. sum(Probabilities parameter) = 1.000018166, but should be 1" thrown in the test body.
[  FAILED  ] ProbDistributionsCategorical.mix_t (0 ms)
[----------] 1 test from ProbDistributionsCategorical (0 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (0 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] ProbDistributionsCategorical.mix_t
```

Is there a way to workaround this that you know of?

Reply:
Ah yeah, I forgot about this, since the arguments from the finite difference are effectively unconstrained, you gotta constrain them.

So something like:

```
return [=](const auto& x2) { return stan::math::categorical_log(x1, stan::math::softmax(x2)); };
```

Reply:
And so for one argument the gradients will all be zero (since the softmax output is 1.0 for one argument) -- so maybe it's worth adding 3 argument input tests?

I dunno, length 2 seems pretty good, and long running tests suck.

Reply:
Ah righto, that makes sense. Thanks!